File: SampleCode.py
# This code will get a the custom field values for a contact from Rhythm that has an id of 303353
# then prints out all the values that were received back from Rhythm.

import requests
import json

# gets the custom_field_values for a contact
def getContact(id="303353", authToken="eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCIsImtpZCI6IjVzT1I2MVBMdENFLUxuT1JIQTRxbyJ9.eyJodHRwOi8vcmh5dGhtc29mdHdhcmUuY29tL2N1c3RvbWVyX2lkIjoiaW5jb3NlLm9yZyIsImh0dHA6Ly9yaHl0aG1zb2Z0d2FyZS5jb20vdGVuYW50X2lkIjoiaW5jb3NlLm9yZyIsImlzcyI6Imh0dHBzOi8vaW5jb3NlLW9yZy51cy5hdXRoMC5jb20vIiwic3ViIjoiY1FGaEtSaVJocDJhZ0VERFA3WXpmbW1sZlBBdUR3Z0hAY2xpZW50cyIsImF1ZCI6Imh0dHBzOi8vYXBpLnJoeXRobXNvZnR3YXJlLmNvbSIsImlhdCI6MTY1MDU1NjY1NSwiZXhwIjoxNjUwNjQzMDU1LCJhenAiOiJjUUZoS1JpUmhwMmFnRUREUDdZemZtbWxmUEF1RHdnSCIsImd0eSI6ImNsaWVudC1jcmVkZW50aWFscyJ9.p-oCk8FAqu0Qo9vtwe0rnD6cM3pDkP_OglanNyFu_xNF3xaEQyp_7rS9gAEOBqN6zgMX9MB1fx352EGzJiVcvBSOQZtEpVasU7gGiU4SvLkeAmdVvP0P8w7-hgUhtKd-8As55x-FMGlaeERoA1qccanxxHsL1ztei5bVseGIrT1q3zGbZhHEtTnzh4ttf5ezXGF7J8_5pKJSWWrjzlpSFaRu43GaAhlN2rrOUyc80CllX9vhjO-3-oRu7lz7OxbIYsPIZD09CnFKaD_wFdcJfw0uhfuyCyLdcMzJmSdGwsM9sUvGPWHwKO5nFcI0UZGqa4wo0tCSsoXummQyr4bJVA"):
    tenant_id = "incose.org"
    url = "https://rolodex.api.rhythmsoftware.com/contacts/" + tenant_id + "/" + id

    query = {
    "fields": "custom_field_values",
    }

    headers = {"Authorization": authToken}

    response = requests.get(url, headers=headers, params=query)

    data = response.json()
    print(data)
    return data


# Each custom field value I store JSON in 'string_value' which can later be  
# deserialized into an object.
#
# {
#     "custom_field_values" : [
#         {
#             "custom_field_id" : "rolodex-contacts__seee_competency_questionnaire_results__c",
#             "string_value": '{ ... }'
#         },
#         { ... },
#         { ... }
#     ]
# }
custom_field_values = getContact()
i = 0
for custom_field in custom_field_values["custom_field_values"]:
    custom_field_id = custom_field["custom_field_id"]
    string_value = custom_field["string_value"]
    print(custom_field_id)
    print(string_value)
    print("index = ", i)
    i += 1

File: SampleCode.js
// contactId is the id of the contact you are trying to get or patch
const contactId = "303353";
// authToken is a generated machine to machine token that is only valid for 24 hours (This one has already expired)
const authToken = "eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCIsImtpZCI6IjVzT1I2MVBMdENFLUxuT1JIQTRxbyJ9.eyJodHRwOi8vcmh5dGhtc29mdHdhcmUuY29tL2N1c3RvbWVyX2lkIjoiaW5jb3NlLm9yZyIsImh0dHA6Ly9yaHl0aG1zb2Z0d2FyZS5jb20vdGVuYW50X2lkIjoiaW5jb3NlLm9yZyIsImlzcyI6Imh0dHBzOi8vaW5jb3NlLW9yZy51cy5hdXRoMC5jb20vIiwic3ViIjoiY1FGaEtSaVJocDJhZ0VERFA3WXpmbW1sZlBBdUR3Z0hAY2xpZW50cyIsImF1ZCI6Imh0dHBzOi8vYXBpLnJoeXRobXNvZnR3YXJlLmNvbSIsImlhdCI6MTY1MDM4NTgzNCwiZXhwIjoxNjUwNDcyMjM0LCJhenAiOiJjUUZoS1JpUmhwMmFnRUREUDdZemZtbWxmUEF1RHdnSCIsImd0eSI6ImNsaWVudC1jcmVkZW50aWFscyJ9.IRf32S97cwZxvYTRUP8K1GNoZ1J5Zttrordp9DGETaFpbuSl3M6J-JjyE-WbryTq4F8cDtLMBvxJYJrgpzvQ7Yb630Q5FvzbdPBHCkhCy2_6taRcqrDRLbtOfesnoeu1t0iNhMScAgQKb6u4rqyou8eR0KhEpUhEV19s5PPnFGnxTAnO3IsLOeaM5ZIXYLd7GZfjtbcjSoc7gJNVLwStFE8aekdkHDqAxw4_xFrrX027DOt51ZGXtAQts4mOdYWuKL-aQfaDSrKXTAiMvKrjHww_05px0SMh28I3GfHqeKeN8iCp4G0SCIsS41Qr2opveS4n7TR87GWRfOTomLIqfQ";
// tenantId is the id that is associated with our resources in Rhythm
const tenantId = "incose.org"

// If patchContact isn't defined yet, create the function to update the custom_field_values on the contact with the corresponding contactId
if (typeof patchContact == 'undefined') {
    patchContact = function (cfv, path="/custom_field_values", tenantId, contactId, authToken) {
        return new Promise(async (resolve, reject) => {
            const resp = await fetch(
                `https://rolodex.api.rhythmsoftware.com/contacts/${tenantId}/${contactId}`,
                {
                    method: 'PATCH',
                    headers: {
                        'Content-Type': 'application/json',
                        Authorization: `${authToken}`
                    },
                    body: JSON.stringify([
                        {
                            op: 'replace',
                            path: `${path}`,
                            value: `${cfv}`
                        }
                    ])
                }
            );

            const data = await resp.json();
            console.log(data);
            resolve(data);
        });
    }
}
if (typeof getContact == 'undefined') {
    getContact = function (query = "", tenantId = 'incose.org', contactId, authToken) {
        return new Promise(async (resolve, reject) => {
            const resp = await fetch(
                `https://rolodex.api.rhythmsoftware.com/contacts/${tenantId}/${contactId}?${query}`,
                {
                    method: 'GET',
                    headers: {
                        Authorization: `${authToken}`
                    }
                }
            );

            const data = await resp.text();
            console.log(data);
            resolve(data);
        });
    }
}

const query = new URLSearchParams({
    fields: 'first_name, last_name, email_address, home_address, facebook, notes, custom_field_values'
}).toString();

var customField = null;
var courseDirectory = null;
var courseDirectoryIndex = 0;
var programList = null;
var programListIndex = 0;
const customFieldPrefix = "rolodex-contacts__";
const dirId = "university_course_directory__c";
const listId = "university_program_list__c";
const fieldValueToDelete = "Anthropology";
getContact(query).then(
    (resp) => {
        if (resp) {
            customField = JSON.parse(resp)["custom_field_values"];
            for (let i = 0; i < customField.length; i++) {
                if (customField[i].custom_field_id === `${customFieldPrefix}${dirId}`) {
                    courseDirectoryIndex = i;
                } else if (customField[i].custom_field_id === `${customFieldPrefix}${listId}`) {
                    programListIndex = i;
                } else {
                    // Do nothing
                    continue;
                }
            }
            courseDirectory = JSON.parse(customField[courseDirectoryIndex].string_value);
            programList = JSON.parse(customField[programListIndex].string_value);
        }
        programList.forEach((program, index) => {
            if (program.programName === fieldValueToDelete) {
                if (!customField) {
                    return;
                }
                programList.splice(index, 1).toString();
                customField[programListIndex].string_value = programList;
                console.log(JSON.stringify(customField));
                patchContact(JSON.stringify(customField)).then((value) => {
                    //window.location.reload(true);
                });
            }
        });
    }
);

/*
 * "ProgramList": [
    {
      "programName": "Computer Science",
      "specializationArea": "Artificial Intelligence",
      "programType": "Bachelor's of Science",
      "totalLectureHours": 45,
      "courseList": [
        {
          "id": "CSCI 1410"
        },
        {
          "id": "CSCI 1411"
        },
        {
          "id": "CSCI 1510"
        },
        {
          "id": "CSCI 1511"
        },
        {
          "id": "CSCI 2412"
        }
      ]
    }
  ]
 * 
 * 
 *"custom_field_values" : [
 *  {...},
 *  {
 *      "custom_field_id": "rolodex-contacts__university_program_list__c",
 *      "string_value" : '"ProgramList": [{ "programName": "Computer Science", "specializationArea": "Artificial Intelligence", "programType": "Bachelor's of Science", "totalLectureHours": 45, "courseList": [{ "id": "CSCI 1410" }, { "id": "CSCI 1411" }, { "id": "CSCI 1510" }, { "id": "CSCI 1511" }, { "id": "CSCI 2412" }]}]'
 *  },
 *  {...}
 * ]
 * 
 * 
 * /